diode-server/netboxdiodeplugin/client.go

```diff
@@ -254,6 +257,110 @@ func protoToJSON(proto proto.Message) (json.RawMessage, error) {
 	return json.RawMessage(jsonBytes), nil
 }
 
+// Authenticate authenticates the client
+func (c *Client) Authenticate(ctx context.Context) error {
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.tokenURL, nil)
+	if err != nil {
+		return err
+	}
+
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	body := url.Values{}
+	body.Add("client_id", c.clientID)
+	body.Add("client_secret", c.clientSecret)
+	body.Add("grant_type", "client_credentials")
+	req.Body = io.NopCloser(strings.NewReader(body.Encode()))
+
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if closeErr := resp.Body.Close(); closeErr != nil {
+			c.logger.Warn("failed to close response body", "error", closeErr)
+		}
+	}()
+
+	respBytes, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("unexpected status code: %d, body: %s", resp.StatusCode, respBytes)
+	}
+
+	var tokenResponse struct {
+		Token string `json:"token"`
+	}
+	if err := json.Unmarshal(respBytes, &tokenResponse); err != nil {
+		return err
+	}
+
+	if tokenResponse.Token == "" {
+		return fmt.Errorf("token is required but not provided in response")
+	}
+
+	c.tokenMutex.Lock()
+	c.token = tokenResponse.Token
+	c.tokenMutex.Unlock()
+	return nil
+}
+
+func (c *Client) doRequestWithRetries(ctx context.Context, method, url string, body io.Reader, headers map[string]string, badRequestErrorCode diodeErrors.ErrorCode) ([]byte, error) {
+	for attempt := 0; attempt < c.maxRetries; attempt++ {
+		req, err := http.NewRequestWithContext(ctx, method, url, body)
+		if err != nil {
+			return nil, err
+		}
+
+		for key, value := range headers {
+			req.Header.Set(key, value)
+		}
+
+		resp, err := c.httpClient.Do(req)
+		if err != nil {
+			return nil, err
+		}
+		defer func() {
+			if closeErr := resp.Body.Close(); closeErr != nil {
+				c.logger.Warn("failed to close response body", "error", closeErr)
+			}
+		}()
+
+		respBytes, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read response body: %w", err)
+		}
+
+		c.logger.Debug("Request to netbox", "statusCode", resp.StatusCode, "response", string(respBytes))
+
+		if resp.StatusCode == http.StatusUnauthorized && attempt < c.maxRetries-1 {
+			c.logger.Info("received 401, attempting reauthentication and retry", "attempt", attempt+1)
+			if err := c.Authenticate(ctx); err != nil {
+				return nil, fmt.Errorf("reauthentication failed: %w", err)
+			}
+			continue // retry
+		}
+
+		// return errors with 4xx status code
+		if resp.StatusCode >= http.StatusBadRequest {
+			var jsonMessage json.RawMessage
+			err := json.Unmarshal(respBytes, &jsonMessage)
+			errorDetails := jsonMessage
+			if err != nil {
+				errorDetails = json.RawMessage(`{"response":"` + string(respBytes) + `"}`)
+			}
+			return nil, changeset.NewError("netbox operation failed", badRequestErrorCode, errorDetails)
```

Comment:
This is unexpected change as `netbox operation failed` will be displayed for both diff and apply main error message. We should retain existing messages maybe per `badRequestErrorCode` something similar.